File: src/vfs/path-utils.ts
export function posixNormalize(inputPath: string): string {
  const raw = inputPath.startsWith('/') ? inputPath : `/${inputPath}`;
  const segments: string[] = [];

  for (const seg of raw.split('/')) {
    if (seg === '' || seg === '.') continue;
    if (seg === '..') {
      segments.pop(); // clamp at root — never goes above /
    } else {
      segments.push(seg);
    }
  }

  return '/' + segments.join('/');
}

export function parentOf(normalizedPath: string): string {
  const idx = normalizedPath.lastIndexOf('/');
  return idx <= 0 ? '/' : normalizedPath.slice(0, idx);
}

export function baseName(normalizedPath: string): string {
  const idx = normalizedPath.lastIndexOf('/');
  return normalizedPath.slice(idx + 1);
}

export function isStrictDescendantPath(ancestorPath: string, candidatePath: string): boolean {
  if (candidatePath === ancestorPath) {
    return false;
  }
  if (ancestorPath === '/') {
    return candidatePath.startsWith('/');
  }
  return candidatePath.startsWith(`${ancestorPath}/`);
}

File: python/src/onetool/vfs/path_utils.py
from __future__ import annotations


def posix_normalize(input_path: str) -> str:
    raw = input_path if input_path.startswith("/") else f"/{input_path}"
    segments: list[str] = []

    for segment in raw.split("/"):
        if segment == "" or segment == ".":
            continue
        if segment == "..":
            if segments:
                segments.pop()
            continue
        segments.append(segment)

    return "/" + "/".join(segments)


def parent_of(normalized_path: str) -> str:
    index = normalized_path.rfind("/")
    if index <= 0:
        return "/"
    return normalized_path[:index]


def base_name(normalized_path: str) -> str:
    index = normalized_path.rfind("/")
    return normalized_path[index + 1 :]


def is_strict_descendant_path(ancestor_path: str, candidate_path: str) -> bool:
    if candidate_path == ancestor_path:
        return False
    if ancestor_path == "/":
        return candidate_path.startswith("/")
    return candidate_path.startswith(f"{ancestor_path}/")
